p2p/conn/connection_test.go

```diff
@@ -112,6 +127,81 @@ func TestMConnectionSend(t *testing.T) {
 	assert.False(t, mconn.Send(0x05, []byte("Absorbing Man")), "Send should return false because channel is unknown")
 }
 
+// round returns x rounded to the nearest int64 (non-negative values only).
+func round(x float64) int64 {
+	if _, frac := math.Modf(x); frac >= 0.5 {
+		return int64(math.Ceil(x))
+	}
+	return int64(math.Floor(x))
+}
+
+func TestMConnectionReceiveRate(t *testing.T) {
+	server, client := NetPipe()
+	defer server.Close()
+	defer client.Close()
+
+	// prepare a client connection with callbacks to receive messages
+	receivedCh := make(chan []byte)
+	errorsCh := make(chan interface{})
+	onReceive := func(chID byte, msgBytes []byte) {
+		receivedCh <- msgBytes
+	}
+	onError := func(r interface{}) {
+		errorsCh <- r
+	}
+
+	cnfg := DefaultMConnConfig()
+	cnfg.SendRate = 500_000 // 500 KB/s
+	cnfg.RecvRate = 500_000 // 500 KB/s
+
+	clientConn := createMConnectionWithCallbacksConfigs(client, onReceive, onError, cnfg)
+	err := clientConn.Start()
+	require.Nil(t, err)
+	defer clientConn.Stop() //nolint:errcheck // ignore for tests
+
+	serverConn := createMConnectionWithCallbacksConfigs(server, func(chID byte, msgBytes []byte) {}, func(r interface{}) {}, cnfg)
+	err = serverConn.Start()
+	require.Nil(t, err)
+	defer serverConn.Stop() //nolint:errcheck // ignore for tests
+
+	msgSize := int(cnfg.RecvRate)
```

Comment:
The reason of why `msgSize` equal to `cnfg.RecvRate` would serve the purpose of this test is that  the traffic rate is actually enforced roughly for every 100 ms (with the max traffic rate of `RecvRate/10`), and, roughly speaking, the peak rate is also measured based on the traffic rate observed in each 100 ms intervals. As such,  a message size configured with `cnfg.RecvRate` can still lead to a sudden increase in traffic for a 100 ms interval hence resulting in a higher peak rate than allowed, which is exactly the purpose of this test.  However,  these specifics may not be immediately apparent when examining the test. Therefore, I agree that it would be more sensible to set the message size to a higher value to prevent any potential confusion. Going to update it.

Reply:
Please see https://github.com/celestiaorg/celestia-core/pull/1119/commits/d71e935b5ee1d198e3b9ea4bb7f3c6137e2964ea